plugins/slack_app/app/controllers/slack_app_controller.rb

```diff
@@ -0,0 +1,152 @@
+                                code: params[:code],
+                                redirect_uri: url_for(only_path: false)
+      if @need_to_connect_app
+        @app_token = SlackIdentifier.create!(identifier: resp['access_token']).identifier
+      end
+      SlackIdentifier.create!(user_id: current_user.id, identifier: resp['user_id'])
+
+      redirect_to
+    end
+  end
+
+  def command
+    render json: command_response
+  end
+
+  def interact
+    # Interesting parts of the payload:
+    # {
+    #   "actions"=>[
+    #     {"name"=>"one", "value"=>"one"} <-- which button the user clicked
+    #   ],
+    #   "callback_id"=>"deploy-123456", <-- maps to a deployment
+    #   "user"=>{"id"=>"U0HAGH3AB", "name"=>"ben"}, <-- who clicked
+    #   "action_ts"=>"1468259216.482422",
+    #   "message_ts"=>"1468259019.000006", <-- need this to maybe update the message
+    #   "response_url"=>"https://hooks.slack.com/actions/T0HAGP0J2/58691408951/iHdjN2zjX0dZ5rXyl84Otql7"
+    # }
+    render json: interact_response # This replaces the original message
+  end
+
+  private
+
+  def command_response
+    # Interesting parts of params:
+    # {
+    #   "user_id"=>"U0HAGH3AB", <-- look up Samson user using this
+    #   "command"=>"/deploy",
+    #   "text"=>"foo bar baz", <-- this is what to deploy
+    #   "response_url"=>"https://hooks.slack.com/commands/T0HAGP0J2/58604540277/g0xd4K2KOsgL9zXwR4kEc0eL"
+    #    ^^^^ This is how to respond later, we can also directly render some JSON
+    # }
+    return unknown_user unless @current_user_from_slack
+
+    # Parse the command
+    projectname, branchname, stagename = @payload['text'].match(/([^\s\/]+)(?:\/(\S+))?\s*(?:to\s+(.*))?/).captures
+
+    # Sanity checks
+    begin
+      project = Project.find_by_param! projectname
+    rescue ActiveRecord::RecordNotFound
+      return unknown_project(projectname)
+    end
+
+    return unauthorized_deployer unless @current_user_from_slack.deployer_for?(project)
+
+    begin
+      stage = project.stages.find_by_param!(stagename || 'production')
+    rescue ActiveRecord::RecordNotFound
+      return unknown_stage(project, stagename)
+    end
+
+    deploy_service = DeployService.new(@current_user_from_slack)
+    deploy = deploy_service.deploy!(stage, reference: branchname || 'master')
+    DeployResponseUrl.create! deploy: deploy, response_url: @payload[:response_url]
+
+    SlackMessage.new(deploy).message_body
+  end
+
+  def interact_response
+    # This means someone clicked the "Approve" button in the channel
+    deploy = Deploy.find_by_id @payload['callback_id']
+    return '_(Whoops, this deploy disappeared.)_' unless deploy
+
+    # Is this user connected?
+    return {
+      replace_original: false,
+      text: unknown_user
+    } unless @current_user_from_slack
+
+    # Can this user approve the deployment?
+    return {
+      replace_original: false,
+      text: "You can't approve your own deploys, silly!"
```

Comment:
Funny is nice ... but I'd prefer people understanding why it does not work
:D

On Sun, Jul 17, 2016 at 3:46 PM, Ben Straub <notifications@github.com>
wrote:

> In plugins/slack_app/app/controllers/slack_app_controller.rb
> <https://github.com/zendesk/samson/pull/1130#discussion_r71091383>:
>
> > +
> > +  def interact_response
> > +    # This means someone clicked the "Approve" button in the channel
> > +    deploy = Deploy.find_by_id @payload['callback_id']
> > +    return '_(Whoops, this deploy disappeared.)_' unless deploy
> > +
> > +    # Is this user connected?
> > +    return {
> > +      replace_original: false,
> > +      text: unknown_user
> > +    } unless @current_user_from_slack
> > +
> > +    # Can this user approve the deployment?
> > +    return {
> > +      replace_original: false,
> > +      text: "You can't approve your own deploys, silly!"
>
> Ah, I see. You want to stomp out all the funny business. Fine, have it
> your way. 😀
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1130/files/d008e855296f051e83d429a560e4e921b94fd917#r71091383>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ-y7Nr_BUpdYc0_YiGJBjh96PkKhks5qWrDegaJpZM4JIRHk>
> .
>